Sources/FirebaseFirestore/Firestore+Swift.swift

```diff
@@ -34,6 +34,63 @@ extension Firestore {
   public func collection(_ collectionPath: String) -> CollectionReference {
     swift_firebase.swift_cxx_shims.firebase.firestore.firestore_collection(self, std.string(collectionPath))
   }
+
+  public func runTransaction(_ updateBlock: @escaping (Transaction, UnsafePointer<NSError?>?) -> Any?, completion: @escaping (Any?, Error?) -> Void) {
+    runTransaction(with: nil, block: updateBlock, completion: completion)
+  }
+
+  public func runTransaction(
+    with options: TransactionOptions?,
+    block updateBlock: @escaping (Transaction, UnsafePointer<NSError?>?) -> Any?,
+    completion: @escaping (Any?, Error?) -> Void
+  ) {
+    let context = TransactionContext(updateBlock: updateBlock)
+    let boxed = Unmanaged.passRetained(context as AnyObject)
+    let future = swift_firebase.swift_cxx_shims.firebase.firestore.firestore_run_transaction(
+      self, options ?? .init(), { transaction, pErrorMessage, pvUpdateBlock in
+        let context = Unmanaged<AnyObject>.fromOpaque(pvUpdateBlock!).takeUnretainedValue() as! TransactionContext
+
+        // Instead of trying to relay the generated `NSError` through firebase's `Error` type
+        // and error message string, just store the `NSError` on `context` and access it later.
+        // We then only need to tell firebase if the update block succeeded or failed and can
+        // just not bother setting `pErrorMessage`.
+
+        // It is expected to run `updateBlock` on whatever thread this happens to be. This is
+        // consistent with the behavior of the ObjC API as well.
+
+        // Since we could run `updateBlock` multiple times, we need to take care to reset any
+        // residue from previous runs. That means clearing out this error field.
+        context.error = nil
+
+        withUnsafePointer(to: context.error) { pError in
+          context.result = context.updateBlock(transaction!.pointee, pError)
+        }
+
+        return context.error != nil ? firebase.firestore.kErrorNone : firebase.firestore.kErrorCancelled
+      },
+      boxed.toOpaque()
+    )
+    future.setCompletion({
```

Comment:
Does someone need to hold onto `future` here, or does the Firebase C++ SDK actually keep this alive just by nature of us creating one?

Reply:
No need to keep the `future` alive. Firebase does that internally. This same approach is used elsewhere.